pkg/querier/frontend/worker_frontend_manager.go

```diff
@@ -0,0 +1,163 @@
+package frontend
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/common/httpgrpc/server"
+	"go.uber.org/atomic"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/grpcclient"
+)
+
+var (
+	backoffConfig = util.BackoffConfig{
+		MinBackoff: 50 * time.Millisecond,
+		MaxBackoff: 1 * time.Second,
+	}
+)
+
+type frontendManager struct {
+	server    *server.Server
+	client    FrontendClient
+	clientCfg grpcclient.Config
+
+	log log.Logger
+
+	workerCancels     []context.CancelFunc
+	managerCtx        context.Context
+	managerCancel     context.CancelFunc
```

Comment:
Given we have all the worker cancellations, do we need a manager level one?  Seems superfluous.